runtime/domain.c

```diff
@@ -1178,7 +1209,12 @@ int caml_try_run_on_all_domains_with_spin_work(
     }
   }
 
-  /* domains now know they are part of the STW */
+  /* Domains now know they are part of the STW.
+
+     Note: releasing the lock will not allow new domain to be created
+     in parallel with the rest of the STW section, as new domains
+     follow the protocol of waiting on [all_domains_cond] which is
+     only signalled at the end of the STW section. */
   caml_plat_unlock(&all_domains_lock);
```

Comment:
This comment triggers a concern for me elsewhere. 

I think there might be a deadlock bug with a domain trying to spawn and all the other domains waiting for that child domain. The deadlock I'm worried about is that the parent domain will be waiting on the child domain to start but if no STW section arises (suppose all other domains block on a variable to be changed by the child) then we will deadlock. 

We need to be careful to call `caml_plat_broadcast(&all_domains_cond)` in places where we might not be guaranteed to wake spawning domains later; i.e. `domain_terminate` and `create_domain`. 

I've submitted a PR to your branch: https://github.com/gasche/ocaml/pull/11

Reply:
My concern is unfounded on closer inspection:
- we use broadcast not signal when `stw_leader` is cleared.
- all the blocked domains will only ever be blocked if they saw `stw_leader != 0`; otherwise they are waiting in `caml_plat_wait` for the `all_domains_lock` to re-enter.